connector/jobrunner/channels.py

```diff
@@ -0,0 +1,602 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#     This file is part of connector, an Odoo module.
+#
+#     Author: Stéphane Bidoul <stephane.bidoul@acsone.eu>
+#     Copyright (c) 2015 ACSONE SA/NV (<http://acsone.eu>)
+#
+#     connector is free software: you can redistribute it and/or
+#     modify it under the terms of the GNU Affero General Public License
+#     as published by the Free Software Foundation, either version 3 of
+#     the License, or (at your option) any later version.
+#
+#     connector is distributed in the hope that it will be useful,
+#     but WITHOUT ANY WARRANTY; without even the implied warranty of
+#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#     GNU Affero General Public License for more details.
+#
+#     You should have received a copy of the
+#     GNU Affero General Public License
+#     along with connector.
+#     If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from datetime import datetime
+from heapq import heappush, heappop
+import logging
+from weakref import WeakValueDictionary
+
+from ..queue.job import PENDING, ENQUEUED, STARTED, FAILED, DONE
+NOT_DONE = (PENDING, ENQUEUED, STARTED, FAILED)
+
+_logger = logging.getLogger(__name__)
+
+
+class PriorityQueue:
+    """A priority queue that supports removing arbitrary objects.
+
+    Adding an object already in the queue is a no op.
+    Popping an empty queue returns None.
+
+    >>> q = PriorityQueue()
+    >>> q.add(2)
+    >>> q.add(3)
+    >>> q.add(3)
+    >>> q.add(1)
+    >>> q[0]
+    1
+    >>> len(q)
+    3
+    >>> q.pop()
+    1
+    >>> q.remove(2)
+    >>> len(q)
+    1
+    >>> q[0]
+    3
+    >>> q.pop()
+    3
+    >>> q.pop()
+    >>> q.add(2)
+    >>> q.remove(2)
+    >>> q.add(2)
+    >>> q.pop()
+    2
+    """
+
+    def __init__(self):
+        self._heap = []
+        self._known = set()    # all objects in the heap (including removed)
+        self._removed = set()  # all objects that have been removed
+
+    def __len__(self):
+        return len(self._known) - len(self._removed)
+
+    def __getitem__(self, i):
+        if i != 0:
+            raise IndexError()
+        while True:
+            if not self._heap:
+                raise IndexError()
+            o = self._heap[0]
+            if o in self._removed:
+                o2 = heappop(self._heap)
+                assert o2 == o
+                self._removed.remove(o)
+                self._known.remove(o)
```

Comment:
In that branch we have a removed Item on top of the heap, so we pop it and try again until a non-removed item is found or the heap is empty (in which case we raise IndexError).